src/devices/logitech_g535.c

```diff
@@ -21,21 +86,121 @@ void g535_init(struct device** device)
-    device_g535.send_sidetone                    = &g535_send_sidetone;
+    /// TODO: usagepage and id may not be correct for battery status and inactive timer
+    device_g535.capability_details[CAP_BATTERY_STATUS] = (struct capability_detail) { .usagepage = 0xc, .usageid = 0x1, .interface = 3 };
+    device_g535.capability_details[CAP_INACTIVE_TIME]  = (struct capability_detail) { .usagepage = 0xc, .usageid = 0x1, .interface = 3 };
+
+    device_g535.send_sidetone      = &g535_send_sidetone;
+    device_g535.request_battery    = &g535_request_battery;
+    device_g535.send_inactive_time = &g535_send_inactive_time;
 
     *device = &device_g535;
 }
 
 static int g535_send_sidetone(hid_device* device_handle, uint8_t num)
 {
+    int ret = 0;
+
     num = map(num, 0, 128, 0, 100);
 
-    uint8_t set_sidetone_level[MSG_SIZE] = { 0x11, 0xff, 0x04, 0x1e, num };
+    uint8_t buf[HIDPP_LONG_MESSAGE_LENGTH] = { HIDPP_LONG_MESSAGE, HIDPP_DEVICE_RECEIVER, 0x04, 0x1d, num, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+
+    ret = hid_send_feature_report(device_handle, buf, sizeof(buf) / sizeof(buf[0]));
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = hid_read_timeout(device_handle, buf, HIDPP_LONG_MESSAGE_LENGTH, hsc_device_timeout);
+    if (ret < 0) {
+        return ret;
+    }
+
+    if (ret == 0) {
+        return HSC_READ_TIMEOUT;
+    }
+
+    // Headset offline
+    if (buf[2] == 0xFF) {
+        return BATTERY_UNAVAILABLE;
+    }
+
+    if (buf[4] != num) {
+        return HSC_ERROR;
+    }
+
+    return ret;
+}
+
+// inspired by logitech_g533.c
+static int g535_request_battery(hid_device* device_handle)
+{
+    int ret = 0;
+
+    // request battery voltage
+    uint8_t buf[HIDPP_LONG_MESSAGE_LENGTH] = { HIDPP_LONG_MESSAGE, HIDPP_DEVICE_RECEIVER, 0x05, 0x0d, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+
+    ret = hid_send_feature_report(device_handle, buf, sizeof(buf) / sizeof(buf[0]));
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = hid_read_timeout(device_handle, buf, HIDPP_LONG_MESSAGE_LENGTH, hsc_device_timeout);
+    if (ret < 0) {
+        return ret;
+    }
+
+    if (ret == 0) {
+        return HSC_READ_TIMEOUT;
+    }
+
+    // Headset offline
+    if (buf[2] == 0xFF) {
+        return BATTERY_UNAVAILABLE;
+    }
+
+    // 7th byte is state; 0x01 for idle, 0x03 for charging
+    uint8_t state = buf[6];
+    if (state == 0x03) {
+        return BATTERY_CHARGING;
+    }
+
+    // actual voltage is byte 4 and byte 5 combined together
+    const uint16_t voltage = (buf[4] << 8) | buf[5];
+
+    return (int)(roundf(poly_battery_level(battery_estimate_terms, num_terms, voltage)));
+}
+
+static int g535_send_inactive_time(hid_device* device_handle, uint8_t num)
+{
+    // Accepted values are 0 (never), 1, 2, 5, 10, 15, 30
```

Comment:
Don't know how to limit the given value here, if a value out of these bounds is given, the headset apparently sets the inactive time to "never"

Reply:
You can return return HSC_OUT_OF_BOUNDS; then